java/common/src/main/scala/org/typelevel/otel4s/java/context/Context.scala

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package java.context
+
+import cats.effect.Sync
+import io.opentelemetry.api.trace.{Span => JSpan}
+import io.opentelemetry.context.{Context => JContext}
+import io.opentelemetry.context.ContextKey
+
+/** A context wrapping a context from the Java open telemetry library (referred
+  * to in further documentation as a "Java context").
+  */
+sealed trait Context {
+
+  /** The underlying Java context. */
+  def underlying: JContext
+
+  /** Retrieves the value associated with the given key from the context, if
+    * such a value exists.
+    */
+  def get[A](key: Context.Key[A]): Option[A]
+
+  /** Retrieves the value associated with the given key from the context, if
+    * such a value exists; otherwise, returns the provided default value.
+    */
+  final def getOrElse[A](key: Context.Key[A], default: => A): A =
+    get(key).getOrElse(default)
+
+  /** Creates a copy of this context with the given value associated with the
+    * given key.
+    */
+  def updated[A](key: Context.Key[A], value: A): Context
+
+  /** Maps the underlying Java context. */
+  private[java] def map(f: JContext => JContext): Context
+}
+
+object Context {
+  private[java] object Noop extends Context {
```

Comment:
Both `Noop` and `Wrapped` do similar things, while `Noop` is more specialized. 
But always falling back to `JSpan.getInvalid.storeInContext(root.underlying)` may lead to the lost context details.

Reply:
I believe that's acceptable. the docs for `getInvalid` state that

> An invalid Span is used when tracing is disabled, usually because there is no OpenTelemetry SDK installed.

I believe that losing context is fine (perhaps even desired) when tracing is disabled

Reply:
You are right. 

For some reason, I thought we would always use a Noop context when the JSpan cannot be extracted. In fact, we use Noop context when JSpan **is present** in the JContext and **is invalid**. So, if JSpan does not exist in the JContext, we use 

I replicated the JContext example from the documentation:

```java
Context withCredential = Context.current().with(CRED_KEY, cred);  
withCredential.wrap(new Runnable() {   
  public void run() {       
    readUserRecords(userId, CRED_KEY.get());    
  }  
}).run()
```

and it works in the similar way:
```scala
import cats.effect.{IO, IOApp, IOLocal}
import org.typelevel.otel4s.java.context.Context

object Example extends IOApp.Simple {

  def run: IO[Unit] =
    IOLocal(Context.root).flatMap { local =>
      for {
        key <- Context.Key.unique[IO, String]("user_id")
        _   <- local.update(_.updated(key, "some_id"))
        ctx <- local.get
        _   <- IO.println(ctx.get(key)) // [info] Some(some_id)
      } yield ()
    }
}
```